### ISSUE TITLE
Issues understanding the magnetostatic force calculations

### ISSUE THREAD

**Issue by seesemichaelj** [halbux/sparselizard#6]
I have this setup (see https://gist.github.com/seesemichaelj/5d9841934caf69149b3e12e444e9e7a4) which has two magnets. The rotor is moved in a simulation and the stator is stationary. The rotor's magnetic north is pointing towards (-1,-1) and the stator's magnetic north is pointing towards (1,1). I move the rotor from -0.25 to 0.25 over multiple calculations. When the rotor is left of the stator (negative alphas), I'd expect negative values of `fx` (i.e. `compx(magforce).integrate(rotor, 5)`) as the south end of the rotor is more pointed towards the south end of the stator. When the rotor is to the right, the north ends repel, causing a positive `fx`.

However, the results I'm getting are:
```
./simulate.sh 10
force (job, fx, fy) = (0, 13.3364, 1.70437) N per unit depth
force (job, fx, fy) = (1, -5.05899, 6.92638) N per unit depth
force (job, fx, fy) = (2, 0.476972, 2.67534) N per unit depth
force (job, fx, fy) = (3, -0.522488, 4.66731) N per unit depth
force (job, fx, fy) = (4, 4.77896, 3.8928) N per unit depth
force (job, fx, fy) = (5, -12.5538, -102.135) N per unit depth
force (job, fx, fy) = (6, 15.2661, 39.6412) N per unit depth
force (job, fx, fy) = (7, 9.64048, 2.41586) N per unit depth
force (job, fx, fy) = (8, 7.58341, -7.8231) N per unit depth
force (job, fx, fy) = (9, -0.243011, 0.765114) N per unit depth
```

Here is a photo of the first, left-most (job 0) mesh:
![image](https://user-images.githubusercontent.com/549323/65497264-cb4b7680-de6e-11e9-9808-b76dba5df405.png)

Am I being dumb (very probable) and not understanding this properly or missing something? Could it be that my mesh size is just not small enough?

I really appreciate the help!

**Comment by halbux** [comment]
Hi Mike,

Happy to see you're going for the magnet simulations! :)

Quick answer to your question:
You are truncating your domain too close to the magnet in the -0.25 case, which creates non-existing forces while they should be extremely low in the -0.25 position (try rdom = 0.4 instead).

I have two remarks:

1.  Your force oscillating values when you move the 'rotor' magnet are due to a too coarse mesh: refine it where needed
2. Your actual question: you are truncating too close to the -0.25 position: try rdom = 0.4 instead of 0.3 and you will see that the force (fine mesh used) drops from the unexpected value (too high and wrong direction) you observe (x: 18.5986, y: -3.63486 N) to the realistic (low value given the distance between the magnets) value (x: -1.03602, y: 0.945544).

As a general note the truncation should in theory be at infinity but in practice far enough to avoid perturbing the magnetic field lines (try increasing rdom values and visualize the solution with (-grad(v)).write(air, "h.pos or .vtk for ParaView") ).

As an illustration of the correctness of the calculated forces I attach a picture of the field lines when the magnets are above each other: the force is as expected high and in the repulsive direction (x: +14, y: +41 Newtons)
![Untitled](https://user-images.githubusercontent.com/34739072/65539347-89234300-df09-11e9-88d2-ebb2fa2d25b3.png)

**Comment by halbux** [comment]
Let me also attach a picture of the magnetic field in the rdom = 0.3 situation with a domain truncation too close to the rotor magnet. You can clearly see the impact of the too close truncation. 
Do not hesitate to visualize the fields! Especially since ParaView support is not extended even more (.vtk, .vtu, .pvd).
BTW: an example of electric motor (with sliding interface) should be available I hope before mid october/end october.
![Untitled](https://user-images.githubusercontent.com/34739072/65540050-ed92d200-df0a-11e9-98a7-415d77ec4c8e.png)

**Comment by seesemichaelj** [comment]
Wow, thanks for the thorough feedback! That all makes a lot of sense, and I'm going to give it a try later tonight, but hopefully to catch you before your night ends: how do you tackle the fine meshing problem? What is an appropriate `Mesh.CharacteristicLengthFactor`? Since I'm meshing at each iteration, and I'm going to make these simulations step at small increments, I want to speed up the meshing. I'm new to gmsh, and this is probably just a generic gmsh question, but hopefully you have some tips that might help.

I'll have to send you some beer/coffee money for all the support you've given me! :)

**Comment by seesemichaelj** [comment]
>BTW: an example of electric motor (with sliding interface) should be available I hope before mid october/end october.

Also, awesome! I definitely will be interested in this when it's ready

**Comment by halbux** [comment]
Thanks for the electric motor enthusiasm!

The classical way is to refine until you see no more change (or small enough) in your solution, then you can assume it has converged.

Gmsh is a fast mesher, but Delaunay (triangular meshes) meshing can take some time!

I see three solutions to that:

1. Use the sparselizard high order interpolation feature: use e.g. an interpolation order 3 with phi.setorder(wholedomain, 3); Finite elements with a mesh size h and an interpolation order p converge as h^(p+1), so that even with a coarser mesh  you will have a more accurate solution at order 3. The downside of course is that the computation will take longer since there are more degrees of freedom

2. Use the sparselizard mesher (only for structured quads/triangle shapes). I don t recommend it in you particular case however since you don't want to re-split the geometry into quadrangle and triangle shapes at every rotor magnet position

3. For close enough rotor magnet positions you don't have to remesh: you could just deform and smooth the mesh with a Laplace formulation (I cannot update it in you code for you right now due to a lack of computer) but there are examples online fpr that (e.g. the Fluid structure interaction example)

I would first start with the easy first option with setorder().

Alex

**Comment by halbux** [comment]
A 4th option: in the gmsh GUI there is the option "Refine by splitting" which is very fast. If there is an equivalent command for the .geo file you could mesh rapidly a coarse mesh then refine by splitting.

**Comment by seesemichaelj** [comment]
The 4th option did work with the `-refine` flag. So now in my script I have:
``` shell
gmsh -format msh2 -2 $geoFile > /dev/null
REFINE_COUNTER=0
while [ ${REFINE_COUNTER} -lt ${NUM_REFINEMENTS} ]; do
  gmsh -format msh2 -refine $mshFile > /dev/null
  let REFINE_COUNTER=REFINE_COUNTER+1 
done
```

Every refinement I do does extend the SL computation a bit (as expected), so this works for me I believe!

I'm still getting some off results however (rdom=0.4, refined the mesh 3 times):
```
force (job, fx, fy) = (0, 0.545863, 0.36756) N per unit depth
force (job, fx, fy) = (1, 0.814026, -0.646231) N per unit depth
force (job, fx, fy) = (2, 1.02643, 1.10372) N per unit depth
force (job, fx, fy) = (3, 2.32796, 1.91265) N per unit depth
force (job, fx, fy) = (4, 11.2834, 2.49638) N per unit depth
force (job, fx, fy) = (5, -9.89457, -98.5835) N per unit depth
force (job, fx, fy) = (6, 13.4331, 41.9617) N per unit depth
force (job, fx, fy) = (7, 8.42133, -0.917453) N per unit depth
force (job, fx, fy) = (8, 1.36828, -1.06635) N per unit depth
force (job, fx, fy) = (9, 0.760642, -1.00867) N per unit depth
```

While the forces are no longer oscillating (except when the rotor passes the stator) due to the increased `rdom` and finer mesh, I still am getting a positive `fx` for the `-0.25` case (and most other cases left of the stator).

I'm trying to reproduce the visualization's you've provided, but I seem to be getting something different.

Using the code:
``` cpp
phi.write(wholedomain, WORKING_DIR + "phi" + fileSuffix.str() + ".pos", 2);
norm(-grad(phi)).write(air, WORKING_DIR + "hnorm" + fileSuffix.str() + ".pos", 2);
(-grad(phi)).write(air, WORKING_DIR + "h" + fileSuffix.str() + ".pos");
```

for `hnorm` (and an `rdom = 0.3` for reproduction), i get teh below graph for the rotor magnet, which looks nothing like your picture:
![image](https://user-images.githubusercontent.com/549323/65578585-e4b1f880-df2a-11e9-905f-9543bf59b3fc.png)

FWIW, my `h(wholedomain)` of the rotor looks like:
![image](https://user-images.githubusercontent.com/549323/65579114-efb95880-df2b-11e9-882f-b0101c628419.png)
which just seems off as H in the magnet is the opposite direction than H in the air

I also updated my sparselizard library (I am using a shared library setup, but I don't see why that would cause any issues).

I either don't have the write code to generate the same charts, or something about my environment is different (maybe?) causing different results. I'm thinking its the former? Although, I can't get the negative `fx` for `alpha=-0.25`

Going to keep looking into it

**Comment by seesemichaelj** [comment]
I'm good to close this; I believe I have a trustworthy setup that I'm happy with (for at least the time being). Thanks! Onto shielding!

**Comment by halbux** [comment]
Ok, great!

The magnets have sharp corners, which probably requires extra refinement and creates nonzero artificial forces (due to field concentration at the corner singularities) when they should be zero... this is a usual finite element behavior. Try rounding the corners to avoid needing an extra fine mesh there.

The only thing I plotted was (-grad(phi)).write(air, "h.pos", 2) then I manually changed the min-max bounds in the plot. Your plots do not seem to be different than mine to me!
